runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -106,7 +108,46 @@ void J9::RecognizedCallTransformer::transform(TR::TreeTop* treetop)
       case TR::java_lang_Math_min_L:
          processIntrinsicFunction(treetop, node, TR::lmin);
          break;
+      case TR::java_lang_StringUTF16_toBytes:
+         process_java_lang_StringUTF16_toBytes(treetop, node);
+         break;
       default:
          break;
       }
    }
+
+void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::TreeTop* treetop, TR::Node* node)
+   {
+   TR_J9VMBase* fej9 = static_cast<TR_J9VMBase*>(comp()->fe());
+
+   TR::Node* valueNode = node->getChild(0);
+   TR::Node* offNode = node->getChild(1);
+   TR::Node* lenNode = node->getChild(2);
+
+   anchorAllChildren(node, treetop);
+   prepareToReplaceNode(node);
+
+   int32_t byteArrayType = fej9->getNewArrayTypeFromClass(reinterpret_cast<TR_OpaqueClassBlock*>(fej9->getJ9JITConfig()->javaVM->byteArrayClass));
+
+   TR::Node::recreateWithoutProperties(node, TR::newarray, 2, 
+      TR::Node::create(TR::ishl, 2,
+         lenNode, 
+         TR::Node::iconst(1)),
+      TR::Node::iconst(byteArrayType), 
+
+      getSymRefTab()->findOrCreateNewArraySymbolRef(node->getSymbolReference()->getOwningMethodSymbol(comp())));
+
+   TR::Node* newByteArrayNode = node;
+   newByteArrayNode->setCanSkipZeroInitialization(true);
+   newByteArrayNode->setIsNonNull(true);
+
+   TR::Node* newCallNode = TR::Node::createWithSymRef(node, TR::call, 5, 
+      getSymRefTab()->methodSymRefFromName(comp()->getMethodSymbol(), "java/lang/String", "decompressedArrayCopy", "([CI[BII)V", TR::MethodSymbol::Static));
```

Comment:
In theory I think yes this would be a cause for concern. In practice this fabricated call will get reduced to an arraycopy IL by the JIT in a subsequent optimization pass so the call would never "exist" to codegen per say. 

We can in fact just generate an arraycopy here instead. However the reason this was implemented in this way is to avoid all the complexities of spine checks, etc. for balanced GC for which code already exists elsewhere in the JIT.

We can disable this optimization under AOT if this is concerning. Please advise.

Reply:
I'm fine with this if the call would never "exist" to the codegen in an AOT compile, i.e. are we sure that the subsequent optimization pass would in fact reduce the call to arraycopy under AOT specifically (I ask, since AOT compiles can be conservative in subtle ways and so I thought I would check) ?

Reply:
The reduction happens here. It's nearly a 1000 lines of code to transform a `java.lang.System.arraycopy` call to an arraycopy IL, hence why I didn't want to replicate any of that logic in this simple transformation:

https://github.com/eclipse/omr/blob/master/compiler/optimizer/ValuePropagationCommon.cpp#L1031-L1933

I suppose this begs the question. Under AOT we are now relying on VP running and eliminating the fabricated call introduced by this transformation. Is this valid? i.e. relying on external optimizations running to clean up our mess here. The only solutions I see here are:

1. We leave it as is
2. We disable for AOT
3. We enable for AOT, eliminate the fabricated call and reduce to an arraycop IL right in this transformation, but then we have to disable under arraylets mode (too much complexity!)
4. We implement this transformation in VP where we are sure the fabricated call will get reduced
5. ???

Please advise.

Reply:
An alternate idea: What if we pushed such helpers out of the String class and into something like JITHelpers and force this class to be fully loaded and resolved before any compilations happen? JITHelpers already pretty much only contains methods which are critical for performance and are super internal to the JDK, so I figure forcing resolution of such helpers should be valid.

In such a world we can always use the JITHelpers as fabricated methods under AOT without any worries. Thoughts?

Reply:
What happens if the call is not reduced to arraycopy ? I assume the method implementation will still keep us functionally correct...it will only be a performance problem, correct ?

Reply:
> What happens if the call is not reduced to arraycopy ? I assume the method implementation will still keep us functionally correct...it will only be a performance problem, correct ?

That's correct. We should be functionally ok if the reduction does not happen. The Java implementation will simply copy one char at a time.

Reply:
My opinion is to disable this under AOT for now since it is'nt clear to me how relocation would be done for a fabricated method call in case it was not reduced to arraycopy. I believe @dsouzai is working on a solution to be able to relocate successfully in more cases and I hope he will be able to enable this optimization as well as part of the work he is doing.

Reply:
Fixed in 5600ddb